tool/util/go.go

```diff
@@ -89,3 +91,70 @@ func SplitCompileCmds(input string) []string {
 func IsGoFile(path string) bool {
 	return strings.HasSuffix(strings.ToLower(path), ".go")
 }
+
+// flagsWithPathValues contains flags that accept a directory or file path as value.
+// From: go help build
+//
+//nolint:gochecknoglobals // private lookup table
+var flagsWithPathValues = map[string]bool{
+	"-o":       true,
+	"-modfile": true,
+	"-overlay": true,
+	"-pgo":     true,
+	"-pkgdir":  true,
+}
+
+// GetBuildPackages loads all packages from the go build command arguments.
+// Returns a list of loaded packages. If no package patterns are found in args,
+// defaults to loading the current directory package.
+// The args parameter should be the go build command arguments (e.g., ["build", "-a", "./cmd"]).
+// Returns an error if package loading fails or if invalid patterns are provided.
+// For example:
+//   - args ["build", "-a", "./cmd"] returns packages for "./cmd"
+//   - args ["build", "-a", "cmd"] returns packages for the "cmd" package in the module
+//   - args ["build", "-a", ".", "./cmd"] returns packages for both "." and "./cmd"
+//   - args ["build"] returns packages for "."
+func GetBuildPackages(args []string) ([]*packages.Package, error) {
+	buildPkgs := make([]*packages.Package, 0)
+	cfg := &packages.Config{
+		Mode: packages.NeedName | packages.NeedFiles | packages.NeedModule,
+	}
+	found := false
+	for i := len(args) - 1; i >= 0; i-- {
+		arg := args[i]
+
+		// If preceded by a flag that takes a path value, this is a flag value
+		// We want to avoid scenarios like "go build -o ./tmp ./app" where tmp also contains Go files,
+		// as it would be treated as a package.
+		if i > 0 && flagsWithPathValues[args[i-1]] {
+			break
+		}
+
+		// If we hit a flag, stop. Packages come after all flags
+		// go build [-o output] [build flags] [packages]
+		if strings.HasPrefix(arg, "-") || arg == "go" || arg == "build" || arg == "install" {
+			break
+		}
+
+		pkgs, err := packages.Load(cfg, arg)
```

Comment:
When running `go build -tags pkg main`, we might incorrectly load numerous packages. For instance, we could accidentally load the 'pkg' package.